docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+- facilitate end-to-end ownership of the cli, and in some cases sub command groups with the cli
+- facilitate decentralized development of extensions for the cli
+
+## Anti-goals
+
+Things we want to avoid.
+
+- block on moving existing kubectl commands out of the kubernetes/kubernetes repo
+- rewrite kubectl from the ground up in a new repo
+
+## Non-goals
+
+Even if these are good ideas, don't let them distract us from meeting our goals will simpler solutions.
+
+- build solution for discovering installable plugins and installing them
+  - rely on existing package management solutions for this until we need something more
+- invent new build and distribution infrastructure
+- fix issues with the existing kubectl commands
+- dog fooding the plugin mechanism for core commands
+
+## Reference-level explanation
+
+- build a new cli binary `ktl` (kubectl2) under the `kubernetes/ktl` repo that dispatches to commands developed in other repos
+- keep old cli commands in `kubernetes/kubernetes/cmd/kubectl` and vendor them into `kubectl/ktl`
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
+
+### ktl
+
+In a new repo (`kubernetes/ktl`), create the `ktl` binary that dispatches (both statically and dynamically)
+to commands developed in other repos.
+
+#### Dispatch
+
+Static dispatch:
+
+- vendor in kubernetes/kubernetes/cmd/kubectl cobra commands, and register them under `ktl kube`
+- vendor in kubernetes/kubectl/cmd cobra commands, and register them under `ktl`
+- vendor in commands from other sources as needed over time
+
+Dynamic dispatch:
+
+- support git-style plugins that allow registering new commands under `ktl`
+  - use simplified version of the kubectl plugin implementation - make configuration files optional
+- plugins only purpose is discovery of kubernetes related commands
+  - plugins can leverage shared client cli libraries (whether they are installed as plugins or not)
+- by default throw an error if plugin names conflict with existing commands
+  - this is configurable
+- plugins can be disabled through modifying `~/.ktlconfig`
+
+Overriding existing commands:
+
+- support command alias' to allow overriding one command with another
+  - allows plugins to extend and override built-in commands
+  - allows `ktl kube *` commands to be aliased to `ktl *`
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
+implement an interface that returns a list of cobra (sp13) commands, which ktl can register under the root.
+
+Initially new cli commands can be built in the kubernetes/kubectl, but if needed, development maybe moved to other repos.
```

Comment:
We (openshift) have a significant amount of experience stitching together multiple repos through vendoring.  The experience is sub-par (not a chance I'd do it if they didn't pay me).  In addition, the state of golang as a language makes vendoring without stripping vendor trees difficult.  The more repos you add, the greater the risk of them relying on conflicting levels of base libraries which makes them impractical to vendor.

The simplest example would probably be client-go.  If you have two things you're vendoring that rely on client-go and they don't agree on level, you can't even pass through the basic client config types to both.  Everyone in the entire tree moves together or not at all.  Or ends up trying to write shims through core types like strings.

Reply:
Things are much simpler if you don't strip vendor trees.

Reply:
> Things are much simpler if you don't strip vendor trees.

Eh, they seem so until you start trying to provide infrastructure down the layers.  When you do that, you end up having to partially flatten since second order interfaces (interface with function returning another interface) don't resolve in golang.  Even very simple vendor trees usually require partial flattening since naughty packages like glog use global flag or default http registration that results in panics at runtime.  

Most projects I've seen strip-vendor.

Reply:
Also note that multiple vendor trees will multiply the size of the binary as well.

Reply:
> Also note that multiple vendor trees will multiply the size of the binary as well.

Its not terrible though.  If we are worried about download size, each full kubectl vendor only adds about 5MB.  If we are worried about the install size, I believe it was only about 25-50MB.

Reply:
In general the only vendored package that can be safely left vendored are those that depend only on stdlib.  The odds of a package exposing via its public API a vendored dependency approach unity when more than 2-3 parent packages are imported.  In practice this means that not stripping vendor trees means you can't invoke most of the public methods.

Examples - etcdclient, if we didn't strip gRPC, would be unable to add wrappers or transports or tweak gRPC, because the packages under the vendor are hidden from us.  This is a massive PITA.

I don't think vendoring without stripping is practical except in carefully chosen cases, of which this is not one.